README.md
# MultiprocessingSpider
[[简体中文版]](https://github.com/Xpp521/MultiprocessingSpider/blob/master/README_cn.md "中文版")
## Description
MultiprocessingSpider is a simple and easy to use web crawling and web scraping framework.

## Architecture
![Architecture](https://raw.githubusercontent.com/Xpp521/Images/master/MultiprocessingSpider_Architecture.jpg)

## Dependencies
- requests

## Installation
```
pip install MultiprocessingSpider
```

## Basic Usage
#### MultiprocessingSpider
```python
from MultiprocessingSpider.spiders import MultiprocessingSpider
from MultiprocessingSpider.packages import TaskPackage, ResultPackage


class MyResultPackage(ResultPackage):
    def __init__(self, prop1, prop2, sleep=True):
        super().__init__(sleep)
        self.prop1 = prop1
        self.prop2 = prop2


class MySpider(MultiprocessingSpider):
    start_urls = ['https://www.a.com/page1']

    proxies = [
        {"http": "http://111.111.111.111:80"},
        {"http": "http://123.123.123.123:8080"}
    ]

    def router(self, url):
        return self.parse

    def parse(self, response):
        # Parsing task or new page from "response"
        ...
        # Yield a task package
        yield TaskPackage('https://www.a.com/task1')
        ...
        # Yield a url or a url list
        yield 'https://www.a.com/page2'
        ...
        yield ['https://www.a.com/page3', 'https://www.a.com/page4']

    @classmethod
    def subprocess_handler(cls, package, sleep_time, timeout, retry):
        url = package.url
        # Request "url" and parse data
        ...
        # Return result package
        return MyResultPackage('value1', 'value2')

    @staticmethod
    def process_result_package(package):
        # Processing result package
        if 'value1' == package.prop1:
            return package
        else:
            return None


if __name__ == '__main__':
    s = MySpider()

    # Start the spider
    s.start()

    # Block current process
    s.join()

    # Export results to csv file
    s.to_csv('result.csv')

    # Export results to json file
    s.to_json('result.json')
```
#### FileSpider
```python
from MultiprocessingSpider.spiders import FileSpider
from MultiprocessingSpider.packages import FilePackage


class MySpider(FileSpider):
    start_urls = ['https://www.a.com/page1']

    stream = True

    buffer_size = 1024

    overwrite = False

    def router(self, url):
        return self.parse

    def parse(self, response):
        # Parsing task or new page from "response"
        ...
        # Yield a file package
        yield FilePackage('https://www.a.com/file.png', 'file.png')
        ...
        # Yield a new url or a url list
        yield 'https://www.a.com/page2'
        ...
        yield ['https://www.a.com/page3', 'https://www.a.com/page4']


if __name__ == '__main__':
    s = MySpider()

    # Add a url
    s.add_url('https://www.a.com/page5')

    # Start the spider
    s.start()

    # Block current process
    s.join()
```
#### FileDownloader
```python
from MultiprocessingSpider.spiders import FileDownloader


if __name__ == '__main__':
    d = FileDownloader()

    # Start the downloader
    d.start()
    
    # Add a file
    d.add_file('https://www.a.com/file.png', 'file.png')
    
    # Block current process
    d.join()
```
More examples → [GitHub](https://github.com/Xpp521/MultiprocessingSpider/tree/master/examples "Examples")
### License
[GPLv3.0](https://github.com/Xpp521/MultiprocessingSpider/blob/master/LICENSE.md "License")  
This is a free library, anyone is welcome to modify : )


CHANGES.md
# Release Note
## v1.1.2
#### Refactor
- Remove property "name" from "FileDownloader".
- Complete class "UserAgentGenerator" in "MultiprocessingSpider.Utils".
- Continue to optimize the setter method of each property. An exception will be raised if the value is invalid. "sleep_time" now can be set to 0.
- Change the sleep strategy of subprocess, subprocess will sleep after receiving the task package to prevent multiple requests from being sent at the same time.
___
## v1.1.1
#### Bug Fixes
- Fix "start_urls" invalidation.
___
## v1.1.0
#### Features
- Add overwrite option for "FileSpider".
- Add routing system. After overriding "router" method, you can yield a single url or a url list in your parse method.
#### Bug Fixes
- Fix retry message display error.
#### Refactor
- Optimize setter method. Now you can do this: spider.sleep_time = ' 5'.
- Will not resend request when "status_code" is not between 200 and 300.
##### a) MultiprocessingSpider
- Rename property "handled_url_table" to "handled_urls".
- Remove method "parse", add "example_parse_method".
- "User-Agent" in "web_headers" is now generated randomly.
- Change url_table parsing order, current rule: "FIFP" (first in first parse).
##### b) FileDownloader
- Remove "add_files" method.
___
## v1.0.0
- The first version.

pack.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2020 Xpp521
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
Packing tool.
"""
from os import listdir, remove
from shutil import rmtree, move
from sys import argv, executable
from subprocess import Popen, PIPE
from os.path import join, isfile, isdir, dirname


def exec_cmd(*args):
    """
    Executes a command.
    :param args: the command.
    :return: stdout of the command.
    :raises: RuntimeError.
    """
    res = Popen(args, stdout=PIPE, stderr=PIPE)
    stdout, stderr = res.communicate()
    if 0 != res.returncode:
        raise RuntimeError('Failed to execute <{}> ({}): {}'.format(args, res.returncode, stderr.decode('gbk')))
    else:
        return stdout.decode('gbk')


# The current directory
CUR_DIR = dirname(__file__)

# The root directory
# ROOT = dirname(CUR_DIR)
ROOT = CUR_DIR

# Module name
MODULE_NAME = ''

# Version
VERSION = ''

# Load module name and version
with open(join(ROOT, 'setup.py')) as f:
    for line in f.readlines():
        if line.startswith('NAME'):
            line = line.strip().replace(' ', '').replace('"', '').replace("'", '')
            MODULE_NAME = line.split('=')[-1]
        elif line.startswith('VERSION'):
            line = line.strip().replace(' ', '').replace('"', '').replace("'", '')
            VERSION = line.split('=')[-1]
        if MODULE_NAME and VERSION:
            break


def remove_paths(paths):
    """Remove all directories and files in the given paths."""
    for p in paths:
        try:
            if isfile(p):
                print('Removing file: {}'.format(p))
                remove(p)
            elif isdir(p):
                print('Removing directory: {}'.format(p))
                rmtree(p, True)
        except FileNotFoundError:
            continue


def clean():
    """Remove the old version."""
    remove_paths([join(ROOT, 'build'), join(ROOT, 'dist'), join(ROOT, '{}.egg-info'.format(MODULE_NAME))])


def pack():
    """Package new version."""
    print(exec_cmd(executable, join(ROOT, 'setup.py'), 'sdist', 'bdist_wheel'))
    for p in listdir(CUR_DIR):
        if p in ('build', 'dist', '{}.egg-info'.format(MODULE_NAME)):
            move(join(CUR_DIR, p), join(ROOT, p))


def check():
    """Check the current version."""
    print(exec_cmd('twine', 'check', join(ROOT, 'dist', '*')))


def install():
    """Install the current version."""
    wheel = ''
    for p in listdir('dist'):
        if p.startswith('{}-{}'.format(MODULE_NAME, VERSION)) and p.endswith('.whl'):
            wheel = join('dist', p)
            break
    if wheel:
        print(exec_cmd('pip', 'uninstall', '-y', MODULE_NAME))
        print(exec_cmd('pip', 'install', wheel))
    else:
        raise FileNotFoundError('wheel file does not exist')


def upload():
    """Upload the current version to PYPI."""
    print(exec_cmd('twine', 'upload', join(ROOT, 'dist', '*')))


def main():
    note = '''Packing tool

Usage:
\t1. pack d:\t\tDelete the old version.

\t2. pack p:\t\tPackage new version.

\t3. pack i:\t\tInstall the current version.

\t3. pack u:\t\tUpload the current version.

'''
    if 2 != len(argv):
        print(note)
        return
    operation = argv[1]
    if 'd' == operation:
        clean()
    elif 'p' == operation:
        clean()
        pack()
        check()
    elif 'i' == operation:
        install()
    elif 'u' == operation:
        if 'y' == input('Upload to PYPI (y/n)? '):
            upload()
    else:
        print(note)


if __name__ == '__main__':
    main()


examples/Girl_MeiTuLu.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2020 Xpp521
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
from os import mkdir
from lxml.etree import HTML
from re import search, split
from os.path import join, exists
from MultiprocessingSpider.utils import format_path
from MultiprocessingSpider.spiders import FileSpider
from MultiprocessingSpider.packages import FilePackage


class MeiTuLuSpider(FileSpider):
    """https://www.meitulu.com/"""

    name = 'MeiTuLuSpider'

    proxies = [
        None
    ]

    def router(self, url):
        if url.startswith('https://www.meitulu.com/item/'):
            return self.parse_album
        else:
            return self.parse_other

    def parse_album(self, response):
        html = HTML(response.content.decode('utf8'))
        # 标题
        title = html.xpath('//div[@class="weizhi"]/h1/text()')[0]
        if search(r' \d+/\d+$', title):
            title = title[:title.rfind(' ')]
        # 图片网址
        base_url = html.xpath('//div[@class="content"]/center/img[1]/@src')[0]
        suffix = base_url[base_url.rfind('.'):]
        base_url = base_url[:base_url.rfind('/') + 1]
        div = html.xpath('//div[@class="c_l"]')[0]
        # 发行机构
        source = div.xpath('p[1]/a/text()')[0]
        # 发行机构网址
        source_url = div.xpath('p[1]/a/@href')[0]
        # 图片数量
        num = div.xpath('p[2]/text()')[0]
        if '图片数量' not in num:
            num = div.xpath('p[3]/text()')[0]
        num = int(num[num.find(' ') + 1: num.rfind(' ')])
        # 模特名称
        model = div.xpath('p[last()-1]/a/text()')[0] if div.xpath('p[last()-1]/a') \
            else div.xpath('p[last()-1]/text()')[0]
        model = model[model.find('：') + 1:]
        # 模特网址
        model_url = div.xpath('p[last()-1]/a/@href')
        model_url = model_url[0] if model_url else ''
        # 发行日期
        date = div.xpath('p[last()]/text()')[0]
        # 标签
        tags = ''
        for tag in html.xpath('//div[@class="fenxiang_l"]/a'):
            tags += '\t{}\t{}\r\n'.format(tag.xpath('text()')[0], tag.xpath('@href')[0])
        # 补充说明
        msg = html.xpath('//p[@class="buchongshuoming"]/text()')
        msg = msg[0] if msg else ''
        info = '''{}

模特名称：{}\t\t{}
发行机构：{}\t\t{}
{}
标签：
{}
补充说明：{}'''.format(title, model, model_url, source, source_url, date, tags, msg)
        title = format_path(title)
        if not exists(title):
            mkdir(title)
        with open(join(title, 'info.txt'), 'w', encoding='utf') as file:
            file.write(info)
        for i in range(1, num + 1):
            yield FilePackage('{}{}{}'.format(base_url, i, suffix), '{}{}'.format(i, suffix), title, True)

    def parse_other(self, response):
        yield HTML(response.content.decode('utf8')).xpath('//div[@class="boxs"]/ul/li/a/@href')


def main():
    p_num = input('子进程数量：')
    s_time = input('最大休眠时间（秒）：')
    timeout = 5
    retry = 3
    s = MeiTuLuSpider(p_num, s_time, timeout, retry)
    while True:
        r = s.add_urls(split(r' +', input('网址（多个网址用空格分隔）：').strip()))
        if r:
            s.start()
            s.join()
        else:
            print('格式错误！重新输入。')
            continue


if __name__ == '__main__':
    main()


MultiprocessingSpider/spiders/spiders.py
# -*- coding: utf-8 -*-
#
# Copyright (C) 2020 Xpp521
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
from time import sleep
from os import makedirs
from requests import get
from os.path import exists, join
from random import randint, choice
from multiprocessing import Process, Queue, cpu_count
from ..utils import UserAgentGenerator, try_convert2int
from ..packages import TaskPackage, FilePackage, ResultPackage, SignalPackage


class MultiprocessingSpider:
    # Spider name
    name = 'unset'

    # Proxy ip
    proxies = [None]

    # Request headers
    web_headers = {'User-Agent': UserAgentGenerator().random()}

    # Start urls
    start_urls = []

    def __init__(self, process_num=None, sleep_time=None, timeout=None, retry=None):
        base_class = str(self.__class__.__base__)
        self.__cpu_count = cpu_count()
        self.__base_class = base_class[base_class.find("'") + 1: base_class.rfind("'")]
        # Process pool
        self.__pool = []
        self.retry = retry
        self.timeout = timeout
        self.sleep_time = sleep_time
        self.process_num = process_num
        self.__result = []
        self.__url_table = []
        if self.start_urls:
            self.add_urls(self.start_urls)
        self.__handled_urls = []
        self.__data_queue = Queue()
        self.__result_queue = Queue()

    def router(self, url):
        """Routing table.
        :param url: url.
        :return: parse method for "url".
        """
        raise NotImplementedError

    def example_parse_method(self, response):
        """Parse data from "response".
        ex:

        def parse(self, response):
            ...
            # Yield a task package
            yield TaskPackage('https://www.a.com/page1/task1')
            ...
            # Yield a new url or a url list
            yield 'https://www.a.com/page2'
            ...
            yield ['https://www.a.com/page3', 'https://www.a.com/page4']

        :param response: response object, type: requests.models.Response.
        :rtype: GentileSpider.packages.TaskPackage or url.
        """
        pass

    def add_url(self, url):
        """Add a url.
        :param url: url.
        :rtype: bool.
        """
        handler = self.router(url)
        if isinstance(url, str) and url.startswith('http') and callable(handler):
            self.__url_table.append((url, handler))
            if self.__pool:
                self.__parse()
            return True
        return False

    def add_urls(self, urls):
        """Add urls.
        :param urls: url list.
        :rtype: bool.
        """
        t = []
        if isinstance(urls, (list, tuple)):
            for url in urls:
                if isinstance(url, str) and url.startswith('http'):
                    handler = self.router(url)
                    if callable(handler):
                        t.append((url, handler))
            if t:
                self.__url_table.extend(t)
                if self.__pool:
                    self.__parse()
                return True
        return False

    def __download_web(self):
        """Download web page."""
        i = 0
        while i < len(self.__url_table):
            url_data = self.__url_table[i]
            url = url_data[0]
            if url not in self.__handled_urls:
                n = 0
                while True:
                    if self._retry < n:
                        break
                    try:
                        response = get(url, headers=self.web_headers, timeout=self._timeout)
                    except Exception as e:
                        print('【Web page load failed】{}, message: {}, try re-downloading ({}/{})...'.
                              format(url, e, n, self._retry))
                    else:
                        if 200 <= response.status_code < 300:
                            self.__handled_urls.append(url)
                            yield response, url_data[1]
                        else:
                            print('【Web page load failed】{}, status code: {}'.format(url, response.status_code))
                        break
                    n += 1
                    sleep(randint(0, self._sleep_time))
            i += 1
        self.__url_table.clear()

    def __parse(self):
        """Parse urls in url_table."""
        for response, parse in self.__download_web():
            for p in parse(response):
                if isinstance(p, str):
                    self.add_url(p)
                elif isinstance(p, list):
                    self.add_urls(p)
                elif isinstance(p, TaskPackage):
                    self.__data_queue.put_nowait(p)
            sleep(randint(0, self._sleep_time))

    @classmethod
    def _subprocess_wrapper(cls, data_queue, result_queue, sleep_time, timeout, retry):
        flag = True
        while True:
            p = data_queue.get()
            if SignalPackage.END == p:
                result_queue.put_nowait(p)
                return
            if flag:
                sleep(randint(0, sleep_time))
            if isinstance(p, TaskPackage):
                result_package = cls.subprocess_handler(p, sleep_time, timeout, retry)
                if isinstance(result_package, ResultPackage):
                    result_queue.put_nowait(result_package)
                    flag = result_package.sleep
                else:
                    flag = True

    @classmethod
    def subprocess_handler(cls, package, sleep_time, timeout, retry):
        """Handler function for subprocesses.
        :param package: task package.
        :param sleep_time: max sleep time after each download.
        :param timeout: timeout.
        :param retry: retry count.
        :rtype: MultiprocessingSpider.packages.ResultPackage or its subclass or None.
        """
        raise NotImplementedError

    @staticmethod
    def process_result_package(package):
        """Process result package.
        :return: MultiprocessingSpider.packages.ResultPackage or its subclass or None.
        """
        return package

    def start(self):
        """Start method."""
        self.__start()
        self.__parse()

    def __start(self):
        for _ in range(self._process_num):
            self.__pool.append(Process(target=self._subprocess_wrapper,
                                       args=(self.__data_queue, self.__result_queue,
                                             self._sleep_time, self._timeout, self._retry)))
            self.__pool[-1].start()
        print(self.info)
        print('———————————————Mission (⊙o⊙) Start———————————————')

    def join(self):
        for _ in self.__pool:
            self.__data_queue.put_nowait(SignalPackage.END)
        n = 0
        while True:
            if self._process_num == n:
                break
            result_package = self.__result_queue.get()
            if SignalPackage.END == result_package:
                n += 1
            elif isinstance(result_package, ResultPackage):
                result_package = self.process_result_package(result_package)
                if isinstance(result_package, ResultPackage):
                    json = result_package.json()
                    if json:
                        self.__result.append(json)
        while self.__pool:
            self.__pool.pop().terminate()
        print('———————————————Mission (∩_∩) Accomplished——————————')

    def to_csv(self, path):
        """Export results in csv.
        :param path: csv file path.
        :rtype: bool.
        """
        if self.__result:
            from csv import DictWriter
            with open(path, 'w', encoding='utf', newline='') as f:
                writer = DictWriter(f, self.__result[0].keys())
                writer.writeheader()
                writer.writerows(self.__result)
                return True
        return False

    def to_json(self, path):
        """Export results in json.
        :param path: json file path.
        :rtype: bool.
        """
        if self.__result:
            from json import dump
            dump(self.__result, open(path, 'w', encoding='utf'))
            return True
        return False

    @property
    def info(self):
        return '''\t\tSpider info
name:\t\t\t{}
class:\t\t\t{}
base class:\t\t{}
subprocess count:\t{}
max sleep time:\t\t{}s
timeout:\t\t{}s
retry count:\t\t{}'''.format(self.name, self.__class__.__name__, self.__base_class,
                             self._process_num, self._sleep_time, self._timeout, self._retry)

    @property
    def process_num(self):
        return self._process_num

    @process_num.setter
    def process_num(self, n):
        n = try_convert2int(n, self.__cpu_count)
        if isinstance(n, int) and 1 <= n <= 50:
            self._process_num = n
        else:
            raise ValueError('invalid value')

    @property
    def sleep_time(self):
        return self._sleep_time

    @sleep_time.setter
    def sleep_time(self, t):
        t = try_convert2int(t, 5)
        if isinstance(t, int) and 0 <= t:
            self._sleep_time = t
        else:
            raise ValueError('invalid value')

    @property
    def timeout(self):
        return self._timeout

    @timeout.setter
    def timeout(self, t):
        t = try_convert2int(t, 5)
        if isinstance(t, int) and 0 < t:
            self._timeout = t
        else:
            raise ValueError('invalid value')

    @property
    def retry(self):
        return self._retry

    @retry.setter
    def retry(self, r):
        r = try_convert2int(r, 3)
        if isinstance(r, int) and 0 <= r:
            self._retry = r
        else:
            raise ValueError('invalid value')

    @property
    def url_table(self):
        return self.__url_table

    @property
    def handled_urls(self):
        return self.__handled_urls


class FileSpider(MultiprocessingSpider):
    """File spider base class."""

    # Request headers for subprocesses
    # PS：subprocess will generate "User-Agent" automatically.
    file_headers = {}

    # Download file in chunks
    stream = False

    # Chunk size
    buffer_size = 1024

    # Overwrite file
    overwrite = False

    def router(self, url):
        """Routing table.
        :param url: url.
        :return: parse method for "url".
        """
        raise NotImplementedError

    def example_parse_method(self, response):
        """Parse data from "response".
        ex:

        def parse(self, response):
            ...
            # Yield a file task package
            yield FilePackage(image.png, 'https://www.a.com/image.png')
            ...
            # Yield a new url or a url list
            yield 'https://www.a.com/page2'
            ...
            yield ['https://www.a.com/page3', 'https://www.a.com/page4']

        :param response: response object, type: requests.models.Response.
        :rtype: GentileSpider.packages.FilePackage or url.
        """
        pass

    @classmethod
    def _download_file(cls, url, path, timeout=10, retry=0):
        """
        :param url: file url.
        :param path: storage path.
        :param timeout: timeout.
        :param retry: retry count.
        :rtype: bool.
        """
        headers = cls.file_headers
        headers['User-Agent'] = UserAgentGenerator().random()
        n = 0
        while True:
            if retry < n:
                return False
            try:
                response = get(url, headers=headers, proxies=choice(cls.proxies), timeout=timeout)
            except Exception as e:
                print('【File download failed】{}, message: {}, try re-downloading ({}/{})...'.format(path, e, n, retry))
            else:
                if 200 <= response.status_code < 300:
                    with open(path, 'wb') as file:
                        if cls.stream:
                            for chunk in response.iter_content(cls.buffer_size):
                                if chunk:
                                    file.write(chunk)
                        else:
                            file.write(response.content)
                    print('【File downloaded successfully】{}'.format(path))
                    return True
                else:
                    print('【File download failed】{}, status code: {}'.format(path, response.status_code))
                    return False
            n += 1

    @classmethod
    def subprocess_handler(cls, package, sleep_time, timeout, retry):
        root = package.root
        if root:
            try:
                if not exists(root):
                    makedirs(root)
            except FileExistsError:
                pass
            file_path = join(root, package.name)
        else:
            file_path = package.name
        if exists(file_path) and not cls.overwrite:
            print('【File already exists】{}'.format(file_path))
            return ResultPackage(False)
        else:
            cls._download_file(package.url, file_path, timeout, retry)

    @staticmethod
    def process_data_package(package):
        if isinstance(package, FilePackage):
            return package


class FileDownloader:
    """Multiprocessing file downloader."""
    def __init__(self, process_num=None, sleep_time=None, timeout=None, retry=None,
                 overwrite=False, stream=False, buffer_size=1024, file_headers=None, proxies=None):
        """
        :param process_num: subprocess number.
        :param sleep_time: max sleep time between each download.
        :param timeout: timeout.
        :param retry: retry count.
        :param overwrite: whether to overwrite file.
        :param stream: download file in chunks.
        :param buffer_size: chunk size.
        :param file_headers: request headers.
        :param proxies: proxies dictionary.
        """
        # Process pool
        self.__cpu_count = cpu_count()
        self.__pool = []
        self.retry = retry
        self.stream = stream
        self.proxies = proxies
        self.timeout = timeout
        self.overwrite = overwrite
        self.sleep_time = sleep_time
        self.buffer_size = buffer_size
        self.process_num = process_num
        self.file_headers = file_headers
        self.__data_queue = Queue()
        self.__result_queue = Queue()

    def add_file(self, url, filename, dirname='', https2http=False):
        return self.__data_queue.put_nowait(FilePackage(url, filename, dirname, https2http))

    @classmethod
    def _download_file(cls, url, path, timeout, retry, stream, buffer, headers, proxies):
        headers['User-Agent'] = UserAgentGenerator().random()
        n = 0
        while True:
            if retry < n:
                return False
            try:
                response = get(url, headers=headers, proxies=proxies, timeout=timeout)
            except Exception as e:
                print('【File download failed】{}, message: {}, try re-downloading ({}/{})...'.format(path, e, n, retry))
            else:
                if 200 <= response.status_code < 300:
                    with open(path, 'wb') as file:
                        if stream:
                            for chunk in response.iter_content(buffer):
                                if chunk:
                                    file.write(chunk)
                        else:
                            file.write(response.content)
                    print('【File downloaded successfully】{}'.format(path))
                    return True
                else:
                    print('【File download failed】{}, status code: {}'.format(path, response.status_code))
                    return False
            n += 1

    @classmethod
    def _subprocess_handler(cls, data_queue, result_queue, sleep_time, timeout, retry,
                            overwrite, stream, buffer_size, file_headers, proxies):
        flag = True
        while True:
            p = data_queue.get()
            if SignalPackage.END == p:
                result_queue.put_nowait(p)
                return
            if flag:
                sleep(randint(0, sleep_time))
            if isinstance(p, FilePackage):
                root = p.root
                if root:
                    try:
                        if not exists(root):
                            makedirs(root)
                    except FileExistsError:
                        pass
                    file_path = join(root, p.name)
                else:
                    file_path = p.name
                if exists(file_path) and not overwrite:
                    print('【File already exists】{}'.format(file_path))
                    flag = False
                else:
                    cls._download_file(p.url, file_path, timeout, retry,
                                       stream, buffer_size, file_headers, choice(proxies))
                    flag = True

    def start(self):
        for _ in range(self.__process_num):
            self.__pool.append(Process(target=self._subprocess_handler,
                                       args=(self.__data_queue, self.__result_queue, self.__sleep_time, self.__timeout,
                                             self.__retry, self.__overwrite, self.__stream, self.__buffer_size,
                                             self.__file_headers, self.__proxies)))
            self.__pool[-1].start()
        print(self.info)
        print('———————————————Mission (⊙o⊙) Start———————————————')

    def join(self):
        for _ in self.__pool:
            self.__data_queue.put_nowait(SignalPackage.END)
        n = 0
        while True:
            if self.__process_num == n:
                break
            result_package = self.__result_queue.get()
            if SignalPackage.END == result_package:
                n += 1
        while self.__pool:
            self.__pool.pop().terminate()
        print('———————————————Mission (∩_∩) Accomplished——————————')

    @property
    def info(self):
        return '''\t\tDownloader info
subprocess count:\t{}
max sleep time:\t\t{}s
timeout:\t\t{}s
retry count:\t\t{}'''.format(self.__process_num, self.__sleep_time, self.__timeout, self.__retry)

    @property
    def process_num(self):
        return self.__process_num

    @process_num.setter
    def process_num(self, n):
        n = try_convert2int(n, self.__cpu_count)
        if isinstance(n, int) and 1 <= n <= 50:
            self.__process_num = n
        else:
            raise ValueError('invalid value')

    @property
    def sleep_time(self):
        return self.__sleep_time

    @sleep_time.setter
    def sleep_time(self, t):
        t = try_convert2int(t, 5)
        if isinstance(t, int) and 0 <= t:
            self.__sleep_time = t
        else:
            raise ValueError('invalid value')

    @property
    def timeout(self):
        return self.__timeout

    @timeout.setter
    def timeout(self, t):
        t = try_convert2int(t, 5)
        if isinstance(t, int) and 0 < t:
            self.__timeout = t
        else:
            raise ValueError('invalid value')

    @property
    def retry(self):
        return self.__retry

    @retry.setter
    def retry(self, r):
        r = try_convert2int(r, 3)
        if isinstance(r, int) and 0 <= r:
            self.__retry = r
        else:
            raise ValueError('invalid value')

    @property
    def overwrite(self):
        return self.__overwrite

    @overwrite.setter
    def overwrite(self, o):
        self.__overwrite = bool(o)

    @property
    def stream(self):
        return self.__stream

    @stream.setter
    def stream(self, s):
        self.__stream = bool(s)

    @property
    def buffer_size(self):
        return self.__buffer_size

    @buffer_size.setter
    def buffer_size(self, b):
        b = try_convert2int(b, 1024)
        if isinstance(b, int) and 1024 <= b:
            self.__buffer_size = b
        else:
            raise ValueError('invalid value')

    @property
    def file_headers(self):
        return self.__file_headers

    @file_headers.setter
    def file_headers(self, h):
        self.__file_headers = h if isinstance(h, dict) else {}

    @property
    def proxies(self):
        return self.__proxies

    @proxies.setter
    def proxies(self, p):
        self.__proxies = p if isinstance(p, list) and p else [None]
